File: src/viz/charts.py
"""Fimiliar Vis — Reusable chart builder functions.

Each function returns a go.Figure (Plotly) styled with the fimiliar template.
Import src.theme before calling any function (done automatically via app.py).
"""

import plotly.express as px
import plotly.graph_objects as go
from plotly.subplots import make_subplots
import pandas as pd

from src.config import (
    COLORS_CATEGORICAL,
    COLORS_SEQUENTIAL,
    COLOR_ACCENT,
    COLOR_POSITIVE,
    COLOR_NEGATIVE,
    TEXT_SECONDARY,
)


def horizontal_bar(
    df: pd.DataFrame,
    x: str,
    y: str,
    title: str = "",
    text_format: str = ",.0f",
    color: str | None = None,
) -> go.Figure:
    """Sorted horizontal bar chart with direct labels.

    Args:
        df: DataFrame with at least columns x and y.
        x: Numeric column (bar length).
        y: Categorical column (bar labels).
        title: Chart title.
        text_format: d3-format string for bar labels.
        color: Optional column for color encoding.
    """
    sorted_df = df.sort_values(x, ascending=True)
    fig = px.bar(
        sorted_df,
        x=x,
        y=y,
        orientation="h",
        title=title,
        color=color,
        color_discrete_sequence=COLORS_CATEGORICAL,
        template="fimiliar",
    )
    fig.update_traces(
        texttemplate=f"%{{x:{text_format}}}",
        textposition="outside",
    )
    if not color:
        fig.update_traces(marker_color=COLOR_ACCENT)
    fig.update_layout(showlegend=bool(color), yaxis_title="", xaxis_title="")
    return fig


def line_chart(
    df: pd.DataFrame,
    x: str,
    y: str | list[str],
    title: str = "",
    color: str | None = None,
) -> go.Figure:
    """Line chart for time series. Max 5 series recommended.

    Args:
        df: DataFrame.
        x: Column for x-axis (typically a date).
        y: Column(s) for y-axis values.
        title: Chart title.
        color: Optional column for color grouping (long-format data).
    """
    if isinstance(y, list):
        fig = go.Figure()
        for i, col in enumerate(y):
            fig.add_trace(go.Scatter(
                x=df[x],
                y=df[col],
                mode="lines",
                name=col,
                line=dict(color=COLORS_CATEGORICAL[i % len(COLORS_CATEGORICAL)], width=2),
            ))
        fig.update_layout(title=title, template="fimiliar")
    else:
        fig = px.line(
            df,
            x=x,
            y=y,
            color=color,
            title=title,
            color_discrete_sequence=COLORS_CATEGORICAL,
            template="fimiliar",
        )
    fig.update_traces(line_width=2)
    return fig


def scatter(
    df: pd.DataFrame,
    x: str,
    y: str,
    title: str = "",
    size: str | None = None,
    color: str | None = None,
    trendline: str | None = "ols",
) -> go.Figure:
    """Scatter plot with optional trendline and size encoding.

    Args:
        df: DataFrame.
        x: Numeric column for x-axis.
        y: Numeric column for y-axis.
        title: Chart title.
        size: Optional column for point size (3rd dimension).
        color: Optional column for color grouping.
        trendline: Trendline type ("ols", "lowess", or None).
    """
    fig = px.scatter(
        df,
        x=x,
        y=y,
        size=size,
        color=color,
        trendline=trendline,
        title=title,
        color_discrete_sequence=COLORS_CATEGORICAL,
        template="fimiliar",
    )
    fig.update_traces(marker=dict(opacity=0.7))
    return fig


def area_chart(
    df: pd.DataFrame,
    x: str,
    y: str,
    color: str | None = None,
    title: str = "",
    normalize: bool = False,
) -> go.Figure:
    """Stacked area chart for composition over time.

    Args:
        df: DataFrame in long format.
        x: Column for x-axis (typically date).
        y: Numeric column for y-axis.
        color: Column for stacking groups.
        title: Chart title.
        normalize: If True, show as 100% stacked.
    """
    fig = px.area(
        df,
        x=x,
        y=y,
        color=color,
        title=title,
        groupnorm="percent" if normalize else None,
        color_discrete_sequence=COLORS_CATEGORICAL,
        template="fimiliar",
    )
    return fig


def histogram(
    df: pd.DataFrame,
    x: str,
    title: str = "",
    nbins: int = 30,
    show_mean: bool = True,
    show_median: bool = True,
) -> go.Figure:
    """Histogram with optional mean/median reference lines.

    Args:
        df: DataFrame.
        x: Numeric column to distribute.
        title: Chart title.
        nbins: Number of bins.
        show_mean: Show vertical mean line.
        show_median: Show vertical median line.
    """
    fig = px.histogram(
        df,
        x=x,
        nbins=nbins,
        title=title,
        color_discrete_sequence=[COLOR_ACCENT],
        template="fimiliar",
    )
    if show_mean:
        mean_val = df[x].mean()
        fig.add_vline(
            x=mean_val,
            line_dash="dash",
            line_color=TEXT_SECONDARY,
            annotation_text=f"Mean: {mean_val:,.1f}",
            annotation_font_size=11,
        )
    if show_median:
        median_val = df[x].median()
        fig.add_vline(
            x=median_val,
            line_dash="dot",
            line_color=COLORS_CATEGORICAL[2],
            annotation_text=f"Median: {median_val:,.1f}",
            annotation_font_size=11,
        )
    return fig


def donut(
    df: pd.DataFrame,
    values: str,
    names: str,
    title: str = "",
) -> go.Figure:
    """Donut chart — use only for 2-3 categories maximum.

    Args:
        df: DataFrame.
        values: Numeric column for slice sizes.
        names: Categorical column for slice labels.
        title: Chart title.
    """
    fig = px.pie(
        df,
        values=values,
        names=names,
        title=title,
        hole=0.55,
        color_discrete_sequence=COLORS_CATEGORICAL,
        template="fimiliar",
    )
    fig.update_traces(
        textposition="inside",
        textinfo="percent+label",
    )
    return fig


def combo_chart(
    df: pd.DataFrame,
    x: str,
    bar_y: str | list[str],
    line_y: str | list[str],
    title: str = "",
    bar_names: list[str] | None = None,
    line_names: list[str] | None = None,
) -> go.Figure:
    """Stacked subplots: bars on top, lines on bottom, shared x-axis.

    Avoids dual y-axes per CLAUDE.md design rules.
    """
    bar_cols = [bar_y] if isinstance(bar_y, str) else bar_y
    line_cols = [line_y] if isinstance(line_y, str) else line_y
    b_names = bar_names or bar_cols
    l_names = line_names or line_cols

    fig = make_subplots(
        rows=2,
        cols=1,
        shared_xaxes=True,
        vertical_spacing=0.08,
        row_heights=[0.55, 0.45],
    )
    for i, (col, name) in enumerate(zip(bar_cols, b_names)):
        fig.add_trace(
            go.Bar(
                x=df[x],
                y=df[col],
                name=name,
                marker_color=COLORS_CATEGORICAL[i % len(COLORS_CATEGORICAL)],
            ),
            row=1,
            col=1,
        )
    for i, (col, name) in enumerate(zip(line_cols, l_names)):
        fig.add_trace(
            go.Scatter(
                x=df[x],
                y=df[col],
                mode="lines",
                name=name,
                line=dict(
                    color=COLORS_CATEGORICAL[
                        (i + len(bar_cols)) % len(COLORS_CATEGORICAL)
                    ],
                    width=2,
                ),
            ),
            row=2,
            col=1,
        )
    fig.update_layout(title=title, template="fimiliar", barmode="group")
    return fig


def grouped_bar(
    df: pd.DataFrame,
    x: str,
    y: str,
    color: str,
    title: str = "",
    text_format: str = ",.0f",
) -> go.Figure:
    """Grouped bar chart using px.bar with barmode='group'."""
    fig = px.bar(
        df,
        x=x,
        y=y,
        color=color,
        barmode="group",
        title=title,
        color_discrete_sequence=COLORS_CATEGORICAL,
        template="fimiliar",
    )
    fig.update_traces(texttemplate=f"%{{y:{text_format}}}", textposition="outside")
    return fig


def waterfall_chart(
    categories: list[str],
    values: list[float],
    title: str = "",
    measure: list[str] | None = None,
) -> go.Figure:
    """Waterfall chart with brand colors for ICP growth decomposition."""
    if measure is None:
        measure = ["relative"] * len(values)
    fig = go.Figure(
        go.Waterfall(
            x=categories,
            y=values,
            measure=measure,
            increasing_marker_color=COLOR_POSITIVE,
            decreasing_marker_color=COLOR_NEGATIVE,
            totals_marker_color=COLORS_CATEGORICAL[3],
            textposition="outside",
            texttemplate="%{y:+.0f}",
        )
    )
    fig.update_layout(title=title, template="fimiliar", showlegend=False)
    return fig


def funnel_chart(
    stages: list[str],
    values: list[float],
    title: str = "",
) -> go.Figure:
    """Funnel chart with sequential palette."""
    fig = go.Figure(
        go.Funnel(
            y=stages,
            x=values,
            textinfo="value+percent initial",
            marker_color=COLORS_SEQUENTIAL[: len(stages)],
        )
    )
    fig.update_layout(title=title, template="fimiliar")
    return fig

File: site/js/charts.js
// Fimiliar Vis — Reusable Plotly.js chart builders

import { PLOTLY_LAYOUT, PLOTLY_CONFIG, CATEGORICAL, SEQUENTIAL, COLORS } from './config.js';

function layout(overrides = {}) {
  return { ...PLOTLY_LAYOUT, ...overrides };
}

export function horizontalBar(el, data, { x, y, title = '', textFormat = ',.0f' } = {}) {
  const sorted = [...data].sort((a, b) => a[x] - b[x]);
  const trace = {
    type: 'bar',
    orientation: 'h',
    x: sorted.map((d) => d[x]),
    y: sorted.map((d) => d[y]),
    marker: { color: COLORS.accent },
    texttemplate: `%{x:${textFormat}}`,
    textposition: 'outside',
  };
  Plotly.newPlot(el, [trace], layout({ title: { text: title, ...PLOTLY_LAYOUT.title }, yaxis: { ...PLOTLY_LAYOUT.yaxis, automargin: true } }), PLOTLY_CONFIG);
}

export function lineChart(el, data, { x, y, title = '', colors = null } = {}) {
  const yFields = Array.isArray(y) ? y : [y];
  const traces = yFields.map((field, i) => ({
    type: 'scatter',
    mode: 'lines',
    x: data.map((d) => d[x]),
    y: data.map((d) => d[field]),
    name: field,
    line: { color: (colors || CATEGORICAL)[i % CATEGORICAL.length], width: 2 },
  }));
  Plotly.newPlot(el, traces, layout({ title: { text: title, ...PLOTLY_LAYOUT.title } }), PLOTLY_CONFIG);
}

export function areaChart(el, data, { x, y, color, title = '', normalize = false } = {}) {
  const categories = [...new Set(data.map((d) => d[color]))];
  const traces = categories.map((cat, i) => {
    const filtered = data.filter((d) => d[color] === cat);
    return {
      type: 'scatter',
      mode: 'lines',
      fill: 'tonexty',
      stackgroup: 'one',
      groupnorm: normalize ? 'percent' : '',
      x: filtered.map((d) => d[x]),
      y: filtered.map((d) => d[y]),
      name: cat,
      line: { color: CATEGORICAL[i % CATEGORICAL.length] },
    };
  });
  Plotly.newPlot(el, traces, layout({ title: { text: title, ...PLOTLY_LAYOUT.title } }), PLOTLY_CONFIG);
}

export function donutChart(el, data, { values, names, title = '' } = {}) {
  const trace = {
    type: 'pie',
    hole: 0.55,
    values: data.map((d) => d[values]),
    labels: data.map((d) => d[names]),
    marker: { colors: CATEGORICAL },
    textposition: 'inside',
    textinfo: 'percent+label',
  };
  Plotly.newPlot(el, [trace], layout({ title: { text: title, ...PLOTLY_LAYOUT.title }, showlegend: false }), PLOTLY_CONFIG);
}

export function comboChart(el, data, { x, barY, lineY, barNames = null, lineNames = null, title = '' } = {}) {
  const barFields = Array.isArray(barY) ? barY : [barY];
  const lineFields = Array.isArray(lineY) ? lineY : [lineY];
  const bNames = barNames || barFields;
  const lNames = lineNames || lineFields;

  const traces = [];
  barFields.forEach((field, i) => {
    traces.push({
      type: 'bar',
      x: data.map((d) => d[x]),
      y: data.map((d) => d[field]),
      name: bNames[i],
      marker: { color: CATEGORICAL[i % CATEGORICAL.length] },
      yaxis: 'y',
    });
  });
  lineFields.forEach((field, i) => {
    traces.push({
      type: 'scatter',
      mode: 'lines',
      x: data.map((d) => d[x]),
      y: data.map((d) => d[field]),
      name: lNames[i],
      line: { color: CATEGORICAL[(i + barFields.length) % CATEGORICAL.length], width: 2 },
      yaxis: 'y2',
    });
  });

  Plotly.newPlot(el, traces, layout({
    title: { text: title, ...PLOTLY_LAYOUT.title },
    barmode: 'group',
    yaxis2: { overlaying: 'y', side: 'right', gridcolor: 'rgba(0,0,0,0)', tickfont: { size: 12 } },
  }), PLOTLY_CONFIG);
}

export function waterfallChart(el, categories, values, { title = '', measure = null } = {}) {
  const measures = measure || (['absolute', ...Array(values.length - 1).fill('relative')]);
  const trace = {
    type: 'waterfall',
    x: categories,
    y: values,
    measure: measures,
    increasing: { marker: { color: COLORS.positive } },
    decreasing: { marker: { color: COLORS.negative } },
    totals: { marker: { color: CATEGORICAL[3] } },
    textposition: 'outside',
    texttemplate: '%{y:+.0f}',
  };
  Plotly.newPlot(el, [trace], layout({ title: { text: title, ...PLOTLY_LAYOUT.title }, showlegend: false }), PLOTLY_CONFIG);
}

export function funnelChart(el, stages, values, { title = '' } = {}) {
  const trace = {
    type: 'funnel',
    y: stages,
    x: values,
    textinfo: 'value+percent initial',
    marker: { color: SEQUENTIAL.slice(0, stages.length) },
  };
  Plotly.newPlot(el, [trace], layout({ title: { text: title, ...PLOTLY_LAYOUT.title } }), PLOTLY_CONFIG);
}

export function scatterChart(el, data, { x, y, title = '', size = null } = {}) {
  const trace = {
    type: 'scatter',
    mode: 'markers',
    x: data.map((d) => d[x]),
    y: data.map((d) => d[y]),
    marker: {
      color: COLORS.accent,
      opacity: 0.7,
      size: size ? data.map((d) => Math.max(6, Math.sqrt(d[size]) / 2)) : 8,
    },
    text: data.map((d) => `${d[x]}, ${d[y]}`),
  };
  Plotly.newPlot(el, [trace], layout({ title: { text: title, ...PLOTLY_LAYOUT.title }, hovermode: 'closest' }), PLOTLY_CONFIG);
}
